src/validators/WebAuthValidator.sol

```diff
@@ -24,49 +24,79 @@ contract WebAuthValidator is VerifierCaller, IModuleValidator {
-  mapping(string originDomain => mapping(address accountAddress => bytes32)) public lowerKeyHalf;
-  mapping(string originDomain => mapping(address accountAddress => bytes32)) public upperKeyHalf;
+  mapping(string originDomain => mapping(bytes credentialId => mapping(address accountAddress => bytes32 publicKey)))
+    public lowerKeyHalf;
+  mapping(string originDomain => mapping(bytes credentialId => mapping(address accountAddress => bytes32 publicKey)))
+    public upperKeyHalf;
+
+  // so you can check if you are using this passkey on this or related domains
+  mapping(string originDomain => mapping(bytes credentialId => address accountAddress)) public keyExistsOnDomain;
+
+  struct PasskeyId {
+    string domain;
+    bytes credentialId;
+  }
 
   /// @notice Runs on module install
   /// @param data ABI-encoded WebAuthn passkey to add immediately, or empty if not needed
   function onInstall(bytes calldata data) external override {
     if (data.length > 0) {
-      require(addValidationKey(data), "WebAuthValidator: key already exists");
+      (bytes memory credentialId, bytes32[2] memory rawPublicKey, string memory originDomain) = abi.decode(
+        data,
+        (bytes, bytes32[2], string)
+      );
+      require(addValidationKey(credentialId, rawPublicKey, originDomain), "WebAuthValidator: key already exists");
     }
   }
 
   /// @notice Runs on module uninstall
   /// @param data ABI-encoded array of origin domains to remove keys for
   function onUninstall(bytes calldata data) external override {
-    string[] memory domains = abi.decode(data, (string[]));
-    for (uint256 i = 0; i < domains.length; i++) {
-      string memory domain = domains[i];
-      lowerKeyHalf[domain][msg.sender] = 0x0;
-      upperKeyHalf[domain][msg.sender] = 0x0;
+    PasskeyId[] memory passkeyIds = abi.decode(data, (PasskeyId[]));
+    for (uint256 i = 0; i < passkeyIds.length; i++) {
+      PasskeyId memory passkeyId = passkeyIds[i];
+      _removeValidationKey(passkeyId.credentialId, passkeyId.domain);
     }
   }
 
-  /// @notice Adds a WebAuthn passkey for the caller
-  /// @param key ABI-encoded WebAuthn public key to add
-  /// @return true if the key was added, false if it was updated
-  function addValidationKey(bytes calldata key) public returns (bool) {
-    (bytes32[2] memory key32, string memory originDomain) = abi.decode(key, (bytes32[2], string));
-    bytes32 initialLowerHalf = lowerKeyHalf[originDomain][msg.sender];
-    bytes32 initialUpperHalf = upperKeyHalf[originDomain][msg.sender];
+  function removeValidationKey(bytes calldata credentialId, string calldata domain) external {
+    return _removeValidationKey(credentialId, domain);
+  }
+
+  function _removeValidationKey(bytes memory credentialId, string memory domain) internal {
+    lowerKeyHalf[domain][credentialId][msg.sender] = 0x0;
+    upperKeyHalf[domain][credentialId][msg.sender] = 0x0;
+    keyExistsOnDomain[domain][credentialId] = address(0);
```

Comment:
Anyone can call onUninstall method and clear this mapping, even if keyExistsOnDomain[domain][credentialId] did not equal to msg.sender before, thus cleaning up data owned by someone else.